engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Vector.java

```diff
@@ -38,23 +42,31 @@ private Vector(Object storage) {
     this.storage = storage;
   }
 
+  public static InvokeFunctionNode createValueForIndexNode() {
+    return InvokeFunctionNode.build(
+        new CallArgumentInfo[] {new CallArgumentInfo()},
+        InvokeCallableNode.DefaultsExecutionMode.EXECUTE,
+        InvokeCallableNode.ArgumentsExecutionMode.EXECUTE);
+  }
+
   @Builtin.Method(
       name = "new",
       description = "Creates new Vector with given length and provided elements.",
       autoRegister = false)
-  @Builtin.Specialize
-  public static Object newFromFunction(long length, Function fun, InteropLibrary interop) {
+  @Builtin.Specialize()
+  public static Object newFromFunction(
+      VirtualFrame frame,
+      long length,
+      Function fun,
+      State state,
+      @Cached("createValueForIndexNode()") InvokeFunctionNode invokeFunctionNode) {
     Object[] target = new Object[Math.toIntExact(length)];
     for (int i = 0; i < target.length; i++) {
-      try {
-        final Object value = interop.execute(fun, (long) i);
```

Comment:
This means that whenever we call into polyglot, then a _callback looses `State`_ - is that OK, @kustosz?